File: financewise_flutter/lib/models/budget.dart
class Budget {
  final int id;
  final int categoryId;
  final double amount;
  final double spent;
  final double remaining;
  final double percentage;
  final String period;
  final DateTime? startDate;
  final DateTime? endDate;
  final bool isActive;

  Budget({
    required this.id,
    required this.categoryId,
    required this.amount,
    this.spent = 0,
    this.remaining = 0,
    this.percentage = 0,
    required this.period,
    this.startDate,
    this.endDate,
    this.isActive = true,
  });

  factory Budget.fromJson(Map<String, dynamic> json) {
    return Budget(
      id: json['id'] ?? 0,
      categoryId: json['category_id'] ?? 0,
      amount: double.tryParse((json['amount'] ?? 0).toString()) ?? 0,
      spent: double.tryParse((json['spent'] ?? 0).toString()) ?? 0,
      remaining: double.tryParse((json['remaining'] ?? 0).toString()) ?? 0,
      percentage: double.tryParse((json['percentage'] ?? 0).toString()) ?? 0,
      period: json['period'] ?? 'monthly',
      startDate: json['start_date'] != null
          ? DateTime.tryParse(json['start_date'].toString())
          : null,
      endDate: json['end_date'] != null
          ? DateTime.tryParse(json['end_date'].toString())
          : null,
      isActive: json['is_active'] ?? true,
    );
  }
}

File: financewise_api/app/Models/Budget.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Budget extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'category_id',
        'amount',
        'period',
        'start_date',
        'end_date',
        'spent',
        'is_active',
    ];

    protected function casts(): array
    {
        return [
            'amount' => 'decimal:2',
            'spent' => 'decimal:2',
            'start_date' => 'date',
            'end_date' => 'date',
            'is_active' => 'boolean',
        ];
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function transactions()
    {
        return $this->hasManyThrough(Transaction::class, Category::class, 'id', 'category_id')
            ->where('type', 'expense')
            ->whereBetween('transaction_date', [$this->start_date, $this->end_date]);
    }

    public function scopeActive($query)
    {
        return $query->where('is_active', true);
    }

    public function getRemainingAttribute()
    {
        return $this->amount - $this->spent;
    }

    public function getPercentageAttribute()
    {
        if ($this->amount == 0) return 0;
        return min(100, round(($this->spent / $this->amount) * 100, 2));
    }

    public function updateSpent()
    {
        $this->spent = $this->transactions()->sum('amount');
        $this->save();
    }
}
